src/asciidoc/docs/guides-general/build-custom-controls.adoc

```diff
@@ -0,0 +1,210 @@
+      setTheme(theme);
+    });
+  }, []);
+  */
+
+  // Listen for changes to the custom control's current state.
+  useState(() => {
+    myCustomCtrl.onUpdate((myProps) => {  // <3>
+      // alert( "myProps.value is " + myProps.value );  // <7>
+    });
+  });
+
+  // When the custom control has loaded, register it to trigger events from the consumer app.
+  useEffect(() => {
+    myCustomCtrl.register();  // <4>
+  }, []);
+
+  // Call the ready method once (with an empty dependency array)
+  // to let the platform know we have our listeners set up and
+  // to unblock the rendering and lookups for the rest of our application.
+  useEffect(() => {
+    Koji.ready();
+  }, []);
+
+  return (
+    <div className="App">
+      <header className="App-header">
+        <h1>My Custom Control</h1>
+        <p>Enter any text:</p>
+        <input
+          onChange={(e) => {
+            myStr = e.currentTarget.value;
+          }}
+        />
+        <button onClick={finish}>{'Save'}</button>
+      </header>
+    </div>
+  );
+}
+
+export default App;
+
+----
+
+[.callout1]#Imports the `CustomVcc` class from the @withkoji/custom-vcc-sdk package.# +
+[.callout2]#Creates an instance of `CustomVcc`.# +
+[.callout3]#Starts event listeners to listen for updates.# +
+[.callout4]#After the customer control loads, registers it to trigger events from the parent editor.# +
+[.callout5]#When a user changes the value, updates the custom control.# +
+[.callout6]#Saves the file.# +
+[.callout7]#For testing. Displays the value returned by the custom control.#
+
+To customize the forked code for your custom control:
+
+. Modify any of the default settings and parameters.
+. Add code required by your custom control, such as:
++
+* Elements in the `return` section that users can interact with.
+* Commands in the `finish` section to perform additional processing on the captured input.
+. Clean up unneeded code.
+. When done, publish your new custom control app.
+
+Learn more: <<customvcc-reference#>>
+
+==== Provide a link to your custom control
+
+Any app that uses your custom control must call `Koji.ui.capture.custom` and provide a `koji-vccs.com` subdomain or URL that points to the custom control.
+
+* You can determine the assigned URL of the custom control.
++
+.Learn how ...
+[%collapsible]
+====
+include::{includespath}/steps-codeeditor.adoc[tag=go-to-project-details-base]
+include::{includespath}/steps-codeeditor.adoc[tag=view-published-koji]
+. Note the URL of the displayed page.
+====
+
+* You can get your own subdomain of `koji-vccs.com`.
++
+.Learn how ...
+[%collapsible]
+====
+include::{includespath}/steps-domains.adoc[tag=open-published-custom-control]
++
+include::{includespath}/steps-domains.adoc[tag=open-published-how]
+
+include::{includespath}/steps-domains.adoc[tag=menu-app-custom-domain,click-add-new-domain,select-subdomain]
+
+include::{includespath}/steps-domains.adoc[tag=create-koji-vccs-com-subdomain]
```

Comment:
Something wonky must be going on in the include file because several steps in this procedure are duplicated. Also, why are there so many includes? It's very hard to maintain, debug, and review when the content is constructed from many tiny pieces. Unless the reuse value is very high, consider writing the text directly in the page.

Reply:
If something that could be reusable is not in an include file, the next person writing similar instructions in another topic would write the same step in their own way, which makes it harder to search and replace later on.

Even if Writer #2 creates an include at that point in time, someone will have to look for other instances of that step everywhere else to convert them into the new include. Again, an issue with search and replace. If they miss any, there would be multiple versions of the same step.

At a previous job, I have actually seen three instances of a piece of information that should be the same, but they all say different things. That feature changed many times and not all references to it were updated. The developer and the PM already left the company, and nobody knew which of those three is the correct information.

I'll look into compressing the steps. There are times when I want the full explanation and others when I just want the summary so I split up the step.

Reply:
Found the cause of the duplicates: one of the end tags didn't match the start tag.

Also, there's now only one step for every screen that they see.